PIL/GifImagePlugin.py

```diff
@@ -552,9 +550,28 @@ def _save_netpbm(im, fp, filename):
 # --------------------------------------------------------------------
 # GIF utilities
 
+# Force optimization so that we can test performance against
+# cases where it took lots of memory and time previously.
+_FORCE_OPTIMIZE = False
+
 def _get_optimize(im, info):
-    return im.mode in ("P", "L") and info and info.get("optimize", 0)
+    if im.mode in ("P", "L") and info and info.get("optimize", 0):
+        # Potentially expensive operation.
 
+        # The palette saves 3 bytes per color not used, but palette
+        # lengths are restricted to 3*(2**N) bytes. Max saving would
+        # be 768 -> 6 bytes if we went all the way down to 2 colors.
+        # * If we're over 128 colors, we can't save any space.
+        # * If there aren't any holes, it's not worth collapsing.
+        # * If we have a 'large' image, the palette is in the noise.
+
+        # create the new palette if not every color is used
+        used_palette_colors = _get_used_palette_colors(im)
+        if _FORCE_OPTIMIZE or im.mode == 'L' or \
+           (len(used_palette_colors) <= 128 and
+            max(used_palette_colors) > len(used_palette_colors) and
```

Comment:
It is really hard to read the code in the if clause with such indention.
Do we have any formatting policy for this case?